source/MaterialXGraphEditor/Graph.cpp

```diff
@@ -694,92 +699,124 @@ void Graph::setRenderMaterial(UiNodePtr node)
+                        if (downstreamNode)
                         {
-                            if (shaderOut[0]->getNode()->getType() == "material")
+                            mx::NodeDefPtr nodeDef = downstreamNode->getNodeDef();
+                            if (nodeDef)
                             {
-                                if (_currRenderNode != shaderOut[0])
+                                if (renderableTypes.count(nodeDef->getType()))
                                 {
-                                    _currRenderNode = shaderOut[0];
-                                    _frameCount = ImGui::GetFrameCount();
-                                    _renderer->setMaterialCompilation(true);
+                                    foundNode = downstreamNode;
+                                    break;
                                 }
                             }
                         }
+                        if (!foundNode)
+                        {
+                            nextPaths.insert(parent->getNamePath());
+                        }
                     }
-                    else
-                    {
-                        _currRenderNode = nullptr;
-                    }
                 }
-                else if (outNodes[0]->getNode()->getType() == mx::MATERIAL_TYPE_STRING)
+                if (foundNode)
+                {
+                    break;
+                }
+            }
+
+            // Set up next set of nodes to search downstream
+            testPaths = nextPaths;
+        }
+
+        // Update rendering. If found use that node, otherwise
+        // use the current fallback of using the first renderable node.
+        if (foundNode)
+        {
+            for (auto uiNode : _graphNodes)
+            {
+                if (uiNode->getNode() == foundNode)
                 {
-                    if (_currRenderNode != outNodes[0])
+                    if (_currRenderNode != uiNode)
                     {
-                        _currRenderNode = outNodes[0];
+                        _currRenderNode = uiNode;
                         _frameCount = ImGui::GetFrameCount();
                         _renderer->setMaterialCompilation(true);
                     }
+                    break;
                 }
             }
-            else
-            {
-                _currRenderNode = nullptr;
-            }
         }
         else
         {
             _currRenderNode = nullptr;
+            _frameCount = ImGui::GetFrameCount();
+            _renderer->setMaterialCompilation(true);
         }
     }
 }
 
 void Graph::updateMaterials(mx::InputPtr input, mx::ValuePtr value)
 {
     std::string renderablePath;
-    mx::TypedElementPtr renderableElem;
-    std::vector<mx::TypedElementPtr> elems = mx::findRenderableElements(_graphDoc);
-
-    size_t num = 0;
-    int num2 = 0;
-    for (mx::TypedElementPtr elem : elems)
+    if (_currRenderNode)
```

Comment:
Simplify this logic as the current node is already found and is already known to be renderable so no need to do a search for renderables and then match.